algo/src/main/java/com/hartwig/serve/sources/ckb/CkbExtractor.java

```diff
@@ -85,33 +81,40 @@ public CkbExtractor(@NotNull Knowledgebase source, @NotNull final EventExtractor
 
     @NotNull
     public ExtractionResult extract(@NotNull List<CkbEntry> entries) {
-        List<ExtractionResult> extractions = Collections.synchronizedList(Lists.newArrayList());
 
         ProgressTracker tracker = new ProgressTracker("CKB", entries.size());
         // Assume entries without variants are filtered out prior to extraction
-        entries.parallelStream().forEach(entry -> { // parallel stream to parallelize transvar calls.
-            if (entry.variants().isEmpty()) {
-                throw new IllegalStateException("A CKB entry without variants has been provided for extraction: " + entry);
-            }
-            int variantCount = entry.variants().size();
-            Variant variant = entry.variants().get(0);
-            String event = variantCount > 1 ? concat(entry.variants()) : CkbEventAndGeneExtractor.extractEvent(variant);
-            String gene = variantCount > 1 ? "Multiple" : CkbEventAndGeneExtractor.extractGene(variant);
-            if (entry.type() == EventType.UNKNOWN) {
-                LOGGER.warn("No event type known for '{}' on '{}'", event, gene);
-            } else {
-                var extraction = getExtractionResult(entry, gene, event, extractions, variant);
-                extractions.add(CkbVariantAnnotator.annotate(extraction, variant));
-            }
-            tracker.update();
-        });
+        List<ExtractionResult> extractions = entries.parallelStream().map(entry -> getExtractionResult(entry, tracker))
+                .flatMap(Optional::stream)
+                .collect(Collectors.toList());
 
         return ExtractionFunctions.merge(extractions);
     }
 
     @NotNull
-    private ExtractionResult getExtractionResult(CkbEntry entry, String gene, String event, List<ExtractionResult> extractions,
-            Variant variant) {
+    private Optional<ExtractionResult> getExtractionResult(@NotNull CkbEntry entry, @NotNull ProgressTracker tracker) {
+        if (entry.variants().isEmpty()) {
+            throw new IllegalStateException("A CKB entry without variants has been provided for extraction: " + entry);
+        }
+        int variantCount = entry.variants().size();
+        Variant variant = entry.variants().get(0);
+        String event = variantCount > 1 ? concat(entry.variants()) : CkbEventAndGeneExtractor.extractEvent(variant);
+        String gene = variantCount > 1 ? "Multiple" : CkbEventAndGeneExtractor.extractGene(variant);
+        
+        Optional<ExtractionResult> result;
+        if (entry.type() == EventType.UNKNOWN) {
+            LOGGER.warn("No event type known for '{}' on '{}'", event, gene);
+            result = Optional.empty();
+        } else {
+            ExtractionResult extraction = extractAndInterpretEvent(entry, gene, event);
+            result = Optional.of(generateKnownEvents ? CkbVariantAnnotator.annotate(extraction, variant) : extraction);
```

Comment:
If we inlined the `extractAndInterpretEvent` maybe we can save some processing when flag is false and remove the second conditional checking generateKnownEvents?

Reply:
I extracted that function just because I thought this one was getting a bit long, but happy to refactor. I will say that in general I think we could find more efficient ways to accomplish CKB extraction as there are many places where we replace certain fields with new information that could probably be streamlined. For example, see `CkbVariantAnnotator.annotate` and `ExtractionFunctions.merge`.

Reply:
My point was that `extractAndInterpretEvent` is only called in this one place, and it is the only caller of `toExtractionResult`. So if we only call it when `generateKnownEvents` is true, then you don't need to check that whether `generateKnownEvents` is true on line 183.

Reply:
`extractAndInterpretEvent` is called for every case where `entry.type() != EventType.UNKNOWN`, so we can't make any assumptions about `generateKnownEvents`.

Reply:
I'm a bit confused, looking at the latest changes I think we missed each other. I meant to inline the call to `extractAndInterpretEvent` such that it would only be called if and when the generateKnownEvents was true. I guess I should have said "inline the variable `extraction`" and not the method... 

That said, if you like the way it looks now I think it still solves my concern, the flag is only checked in one place.

Reply:
I still don't follow your suggestion...

We needed to call `extractAndInterpretEvent` when `generateKnownEvents` was false too. In that case, the `ExtractionResult` was returned without annotation. That said, annotation now happens with extraction so the extra check is no longer needed.

Reply:
I see it now, I'd erroneously read the code as returning Optional.empty when the flag was false. My bad! I think its still nice now though that we don't need to check the flag twice. Alls well that ends well, sorry for any wasted time.

Reply:
No problem, I still think this was an improvement.